### PR TITLE
Fix race in DeltaFIFO

### LINKED ISSUE #31981: [k8s.io] Load capacity [Feature:Performance] should be able to handle 30 pods per node {Kubernetes e2e suite}
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubernetes-kubemark-500-gce/5489/

Failed: [k8s.io] Load capacity [Feature:Performance] should be able to handle 30 pods per node {Kubernetes e2e suite}

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/load.go:198
scaling rc load-big-rc-2 for the first time
Expected error:
    <*errors.errorString | 0xc839362ba0>: {
        s: "error while scaling RC load-big-rc-2 to 219 replicas: timed out waiting for \"load-big-rc-2\" to be synced",
    }
    error while scaling RC load-big-rc-2 to 219 replicas: timed out waiting for "load-big-rc-2" to be synced
not to have occurred
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/load.go:319
```

Previous issues for this test: #26544 #26938 #27595 #30146 #30469 #31374 #31427 #31433 #31589

### PR DESCRIPTION
**Pull request by wojtek-t**
Fix #31981

@ncdc @deads2k @pmorie @gmarek

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32125)

### CODE REVIEW

**Comment by wojtek-t** [comment]
@kubernetes/sig-api-machinery @smarterclayton - FYI

File: pkg/client/cache/delta_fifo.go [hunk 1]
```diff
@@ -522,6 +522,23 @@ func (f *DeltaFIFO) syncKey(key string) error {
 		return nil
 	}
 
+	// If we are doing Resync() and there is already an event queued for that object,
```
Comment (smarterclayton) [inline pkg/client/cache/delta_fifo.go]: Could we pass "sync bool" queueActionLocked()?

Comment (wojtek-t) [inline pkg/client/cache/delta_fifo.go]: Ideally, I think that we should introduce a new type of event "Replace" (next to Added, Updated, Deleted and Sync), but maybe not in 1.4 timeframe?

WDYT?

Comment (smarterclayton) [inline pkg/client/cache/delta_fifo.go]: Hrm - isn't replace Sync?

Comment (wojtek-t) [inline pkg/client/cache/delta_fifo.go]: It is treated as Sync now. But those two are different in fact:
- Sync is supposed to be purely "process my with whatever value I have now"
- Replace is "replace me with that value no matter what value I had before"
Conceptually they are different, but in the code they are treated the same way.

Comment (smarterclayton) [inline pkg/client/cache/delta_fifo.go]: I'm worried about all the places that will now get a new event type that didn't before.

Comment (wojtek-t) [inline pkg/client/cache/delta_fifo.go]: I think there is only one such place (except from DeltaFIFO). But that's the main reason why I don't want to change it for 1.4.

Comment (lavalamp) [inline pkg/client/cache/delta_fifo.go]: "replace" should not be a new event type, just compute the corresponding add/update/delete object.

File: pkg/client/cache/delta_fifo.go [hunk 2]
```diff
@@ -522,6 +522,23 @@ func (f *DeltaFIFO) syncKey(key string) error {
...
+	// we ignore the Resync for it. This is to avoid the race, in which the resync
+	// comes with the previous value of object (since queueing an event for the object
+	// doesn't trigger changing the underlying store <knownObjects>.
+	//
+	// TODO: Ideally this should be handled in queueActionLocked, but currently there
```
Comment (lavalamp) [inline pkg/client/cache/delta_fifo.go]: I don't agree with this TODO, this code looks right to me?

File: pkg/client/cache/delta_fifo.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/client/cache/delta_fifo.go]: I don't understand your comment. Basically we are calling Replace e.g. when we are doing "relist", if we are not able to get all the deltas from in-between. This is effectively an "update", but this is handled in the code as "Sync".

Comment (lavalamp) [inline pkg/client/cache/delta_fifo.go]: That's probably wrong. If it changes a value "sync" is probably wrong. Scary to change, though.

Comment (wojtek-t) [inline pkg/client/cache/delta_fifo.go]: Yeah - we can try changing it in 1.5, but definitely not now.

**Comment by wojtek-t** [comment]
@lavalamp - comment applied, PTAL

**Comment by lavalamp** [comment]
LGTM, let's get this in. Thanks for fixing!!

**Comment by smarterclayton** [comment]
Also LGTM.

**Pull request merged**